gcc/rust/resolve/rust-forever-stack.hxx

```diff
@@ -0,0 +1,227 @@
+{
+  // If the link does not exist, we create it and emplace a new `Node` with the
+  // current node as its parent. `unordered_map::emplace` returns a pair with
+  // the iterator and a boolean. If the value already exists, the iterator
+  // points to it. Otherwise, it points to the newly emplaced value, so we can
+  // just update our cursor().
+  auto emplace
+    = cursor ().children.emplace (std::make_pair (link, Node (rib, cursor ())));
+
+  auto it = emplace.first;
+  auto didnt_exist = emplace.second;
+
+  rust_debug ("inserting link: Link(%d [%s]): existed? %s", link.id,
+	      link.path.has_value () ? link.path.value ().c_str () : "<anon>",
+	      didnt_exist ? "no" : "yes");
+
+  // We update the cursor
+  cursor () = it->second;
+}
+
+template <Namespace N>
+void
+ForeverStack<N>::pop ()
+{
+  assert (!cursor ().is_root ());
+
+  cursor () = cursor ().parent.value ();
+}
+
+template <Namespace N>
+tl::expected<NodeId, AlreadyInserted>
+ForeverStack<N>::insert (Identifier name, NodeId node)
+{
+  auto &innermost_rib = peek ();
+
+  // So what do we do here - if the Rib has already been pushed in an earlier
+  // pass, we might end up in a situation where it is okay to re-add new names.
+  // Do we just ignore that here? Do we keep track of if the Rib is new or not?
+  // should our cursor have info on the current node like "is it newly pushed"?
+  return innermost_rib.insert (name, node);
+}
+
+template <Namespace N>
+Rib &
+ForeverStack<N>::peek ()
+{
+  return cursor ().rib;
+}
+
+template <Namespace N>
+const Rib &
+ForeverStack<N>::peek () const
+{
+  return cursor ().rib;
+}
+
+template <Namespace N>
+void
+ForeverStack<N>::reverse_iter (std::function<KeepGoing (Rib &)> lambda)
+{
+  auto &tmp = cursor ();
+
+  while (true)
+    {
+      auto keep_going = lambda (tmp);
+      if (keep_going == KeepGoing::No)
+	return;
+
+      if (tmp.is_root ())
+	return;
+
+      tmp = tmp.parent.value ();
+    }
+}
+
+template <Namespace N>
+typename ForeverStack<N>::Node &
+ForeverStack<N>::cursor ()
+{
+  return cursor_reference;
+}
+
+template <Namespace N>
+const typename ForeverStack<N>::Node &
+ForeverStack<N>::cursor () const
+{
+  return cursor_reference;
+}
+
+template <Namespace N>
+tl::optional<NodeId>
+ForeverStack<N>::get (const Identifier &name)
+{}
+
+// TODO: Are there different fetching behavior for different namespaces?
+// inline template <>
+// tl::optional<NodeId>
+// ForeverStack<Namespace::Values>::get (const Identifier &name)
+// {
+//   return {};
+// }
+
+template <Namespace N>
+std::string
+ForeverStack<N>::rib_str (const Rib &rib, const std::string &next,
+			  const std::string &next_next)
+{
+  if (rib.get_values ().empty ())
+    {
+      std::cerr << next << "rib: {},\n";
+      return "";
+    }
+
+  std::cerr << next << "rib: {\n";
+
+  for (const auto &kv : rib.get_values ())
+    std::cerr << next_next << kv.first << ": " << kv.second << "\n";
+
+  std::cerr << next << "},\n";
+
+  return "";
```

Comment:
I'm not sure about the intent of this function. Are you dumping the rib to `std::cerr` or a string ? The result string seems kinda useless in the current state. I would construct the string and let the caller send it to `cerr`.

Reply:
yeah, the dump is unfinished - but this is what it should be doing. keep a `stringstream` and then return it, and the session manager can dump it to `std::cerr` or to a file. forgot I hadn't fixed that!